ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullProjectNode.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.planner.plan;
+
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.analyzer.RewrittenAnalysis;
+import io.confluent.ksql.execution.codegen.CodeGenRunner;
+import io.confluent.ksql.execution.codegen.ExpressionMetadata;
+import io.confluent.ksql.execution.plan.SelectExpression;
+import io.confluent.ksql.execution.util.ExpressionTypeManager;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.parser.tree.AllColumns;
+import io.confluent.ksql.parser.tree.SelectItem;
+import io.confluent.ksql.parser.tree.SingleColumn;
+import io.confluent.ksql.planner.Projection;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.LogicalSchema.Builder;
+import io.confluent.ksql.schema.ksql.SystemColumns;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * The projection of a Pull query.
+ *
+ * <p>There are 3 schemas the node is handling, the input schema, the intermediate schema used
+ * for codegen and the output schema.
+ * <ul>
+ * <li>The input is the schema of the child node
```

Comment:
excuse me while I philosophize for a bit...

just wondering, is 'child' and 'parent' standard terminology in databases when discussing logical plans? I know we've standardized on it, but it always confuses me because the "child" is the "source node" which makes me feel like it's a "parent" 🤔 

What do you think about upstream/downstream nodes? That way it's clear that the direction of data is flowing from the source nodes (upstream) to the sink nodes (downstream)

Reply:
The plans (logical, physical) are trees hence the terminology root, child, parent. The children are the leaves and as we move up the tree we visit from child to parent. Information and rows is flowing from the leaves towards the root.

Honestly, sorry but I find the upstream/downstream more confusing :(

Reply:
😂 haha i'll keep working at making it feel intuitive!